### PR TITLE
[hdpowerview] Add support for scene groups

### LINKED ISSUE #11533: [hdpowerview] Support activation of scene collections
The Hunter Douglas PowerView Hub supports two different scene concepts:
- Scenes
- Scene Collections

A scene collection is a collection of scenes and can be used to group multiple scenes and activate them all with a single click or automation.

Currently the binding supports only scenes, but not scene collections.

**Note:** Hunter Douglas has removed scene collections from the app, but they are still supported by the hub. Any user already having scene collections will be able to still trigger them - also from the app. In order to implement #11516 without filtering scene collections, this issue is still relevant, although a bit late.

### PR DESCRIPTION
**Pull request by jlaur**
Fixes #11533

Signed-off-by: Jacob Laursen <jacob-github@vindvejr.dk>

The Hunter Douglas PowerView Hub supports two different scene concepts:
- Scenes
- Scene Collections

A scene collection is a collection of scenes and can be used to group multiple scenes and activate them all with a single click or automation. When this feature was available in the Hunter Douglas PowerView app they were called scene groups. This terminology it kept in user-oriented strings, while API name is scene collections.

This PR adds support for activation of scene groups. Scene groups will appear as channels the same way as scenes, like this:
image

### CODE REVIEW

**Comment by andrewfg** [comment]
I will have a look at this in the next couple of days.

**Comment by andrewfg** [comment]
@jlaur I looked at your code, and FWIW it seems excellent to me; however ..
- I don't use scene collections in my physical system so I could not test it in real life
- I am not an official code reviewer so I cannot formally approve your code either

**Comment by jlaur** [comment]
> @jlaur I looked at your code, and FWIW it seems excellent to me; however ..
> 
> * I don't use scene collections in my physical system so I could not test it in real life
> * I am not an official code reviewer so I cannot formally approve your code either

@andrewfg - thanks for having a look, any additional eyes and feedback is much appreciated no matter who has the stamp. :-) I actually only have a single scene collection as I can't figure out how to create new collections using the PowerView app. It seems that either it's not supported anymore (although I can see and edit the one I already have), or there is a limition of one collection (which would seems strange). Maybe I'll try to create one more using the API directly to at least verify the "limit of one" theory.

**Comment by andrewfg** [comment]
I will do some tests on my hub to try and create such a scene collection.

**Comment by andrewfg** [comment]
Hmm. I just tried the PowerView app on my Android phone (app version dated October 25, 2021). Since I did not have any Scene Collections before, there was nothing that I could edit. And it seems not possible to create a new one now (the app only has tabs for Rooms, Scenes, and Automations). However I will also try on the iPad app, in case it may be different.

**Comment by andrewfg** [comment]
^
The PowerView iPad app latest version (from yesterday) also does not have a tab for creating scene collections.

**Comment by jlaur** [comment]
@andrewfg 
> The PowerView iPad app latest version (from yesterday) also does not have a tab for creating scene collections.

It was the same activity as scenes, but with a context menu for adding "Scene Groups", as they are called in the app. I can still see and activate my scene group using latest version of the app:
image

It could be a bug or a decision to remove the feature. I'll try to reach out to them.

Nevertheless, I think it's still worth having this feature, and it works with the scene group/collection I and other users might already have. Also it provides a kind of backwards compatibility that could be useful for fully supporting #11516 when users are still having some scene collections. Otherwise they would need to be filtered, so IMHO better to have direct support.

I tried to fire up Postman to try to create a new collection manually, but haven't quite figured out how yet.

GET /api/sceneCollections/27119/
```
{
    "sceneCollection": {
        "name": "QsO4cm4gb3A=",
        "colorId": 12,
        "iconId": 17,
        "id": 27119,
        "order": 0,
        "hkAssist": false
    }
}
```

As you can see, this entry doesn't mention the scenesIds it actually contains. So this must be somewhere else - and would be needed for creating a meaningful new collection.

**Comment by andrewfg** [comment]
^
In my case I have no Scene Collection ... and so ( therefore??) there is no context menu either.

**Comment by lolodomo** [comment]
Conflict has to be fixed.

**Comment by jlaur** [comment]
@lolodomo - done, ready again.

**Comment by jlaur** [comment]
@lolodomo - if you are still actively managing pull requests today perhaps you could have a look at this? We lack original maintainer, but @andrewfg has reviewed as another HD PowerView binding contributor.

**Review by andrewfg — APPROVED** [review]
LGTM

**Comment by lolodomo** [comment]
I have two comments, I will push them this evening.

File: bundles/org.openhab.binding.hdpowerview/src/main/resources/OH-INF/i18n/hdpowerview.properties [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+thing-type.config.hdpowerview.shade.id.description = The numeric ID of the PowerView Shade in the Hub
+
+# channel types
+
+channel-type.hdpowerview.battery-voltage.label = Battery Voltage
+channel-type.hdpowerview.battery-voltage.description = Battery voltage reported by the shade
+channel-type.hdpowerview.scene-activate.description = Activates the scene ''{0}''
+channel-type.hdpowerview.scene-group-activate.description = Activates the scene group ''{0}''
+channel-type.hdpowerview.shade-position.label = Position
+channel-type.hdpowerview.shade-position.description = The vertical position of the shade
```
Comment (lolodomo) [inline bundles/org.openhab.binding.hdpowerview/src/main/resources/OH-INF/i18n/hdpowerview.properties:33]: Two entries are missing for label and description of the channel `secondary`.
Keys should be probably `thing-type.hdpowerview.shade.channel.secondary.label` and `thing-type.hdpowerview.shade.channel.secondary.description`.
Please try the result before submitting.

File: bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/HDPowerViewJUnitTests.java [hunk 2]
```diff
@@ -287,80 +286,114 @@ public void testOnlineCommunication() {
...
-            @Nullable
-            String json = loadJson("shades");
-            assertNotNull(json);
-            assertNotEquals("", json);
-            shades = gson.fromJson(json, Shades.class);
-            assertNotNull(shades);
-        } catch (JsonParseException e) {
-            fail(e.getMessage());
-        }
+        @Nullable
```
Comment (lolodomo) [inline bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/HDPowerViewJUnitTests.java:296]: `@Nullable` not expected for local variable.
Please check all the class.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewHubHandler.java [hunk 3]
```diff
@@ -276,8 +297,10 @@ private void pollScenes() throws JsonParseException, HubProcessingException, Hub
             } else {
                 // create a new scene channel
                 ChannelUID channelUID = new ChannelUID(getThing().getUID(), sceneId);
+                String description = translationProvider.getText("channel-type.hdpowerview.scene-activate.description",
+                        scene.getName());
```
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewHubHandler.java]: Good point - fixed.

File: bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewHubHandler.java [hunk 4]
```diff
@@ -292,6 +315,51 @@ private void pollScenes() throws JsonParseException, HubProcessingException, Hub
...
+                logger.debug("Creating new channel for scene collection '{}'", sceneCollectionId);
+            }
+        }
+
+        // remove any previously created channels that no longer exist
+        if (!idChannelMap.isEmpty()) {
+            logger.debug("Removing {} orphan scene collection channels", idChannelMap.size());
+            List<Channel> allChannels = new ArrayList<>(getThing().getChannels());
+            allChannels.removeAll(idChannelMap.values());
+            updateThing(editThing().withChannels(allChannels).build());
```
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewHubHandler.java]: Nice catch, fixed.

File: bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewHubHandler.java [hunk 5]
```diff
@@ -292,6 +315,51 @@ private void pollScenes() throws JsonParseException, HubProcessingException, Hub
...
+            // remove existing scene collection channel from the map
+            String sceneCollectionId = Integer.toString(sceneCollection.id);
+            if (idChannelMap.containsKey(sceneCollectionId)) {
+                idChannelMap.remove(sceneCollectionId);
+                logger.debug("Keeping channel for existing scene collection '{}'", sceneCollectionId);
+            } else {
+                // create a new scene collection channel
+                ChannelUID channelUID = new ChannelUID(getThing().getUID(), sceneCollectionId);
+                String description = translationProvider.getText(
+                        "channel-type.hdpowerview.scene-group-activate.description", sceneCollection.getName());
```
[the remark this answers is not in the document]
Comment (jlaur) [reply bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewHubHandler.java]: Fixed.

File: bundles/org.openhab.binding.hdpowerview/src/main/resources/OH-INF/i18n/hdpowerview.properties (hunk 1 above)
Comment (jlaur) [reply bundles/org.openhab.binding.hdpowerview/src/main/resources/OH-INF/i18n/hdpowerview.properties:33]: Added and verified to be correct. Thanks for this, couldn't find these keys in the documentation. Seems like I missed this commit in the tool for auto-generating: https://github.com/openhab/openhab-core/pull/2544/commits/414c85c9ddcc3311afef86092d70196ab82e32cb

File: bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/HDPowerViewJUnitTests.java (hunk 2 above)
Comment (jlaur) [reply bundles/org.openhab.binding.hdpowerview/src/test/java/org/openhab/binding/hdpowerview/HDPowerViewJUnitTests.java]: Fixed.

**Review by lolodomo — APPROVED** [review]
LGTM

**Comment by andrewfg** [comment]
@jlaur albeit rather late, but I think that the official name for 'Scene Collection' in the HD documentaion is/was 'Multiple Room Scene' -- see below..

image

**Comment by jlaur** [comment]
@andrewfg - do you have a link for that documentation? In the API they are called **scene collections**, while they are called **scene groups** in the app. I decided to be consistent with the app as this is probably the only reference for users, i.e. users have most likely created the scene group within the app and now wants to trigger it from openHAB.

I think what you are referring to is **multiple room scenes**, i.e. scenes including multiple rooms. They are still supported by the app, and it's just a specific kind of scene involving more than one room, so not to be mistaken with scene group. Maybe this feature was added later, thus more or less obsoleted scene groups. Scene groups are still kind of nice IMHO since they can reduce redundancy by having to create scenes for multiple combinations.

**Pull request merged**